clippy_lints/src/paths_from_format.rs

```diff
@@ -0,0 +1,124 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::macros::{root_macro_call, FormatArgsExpn};
+use clippy_utils::sugg::Sugg;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+use std::fmt::Write as _;
+use std::path::Path;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `PathBuf::from(format!(..))` calls.
+    ///
+    /// ### Why is this bad?
+    /// It is not OS-agnostic, and can be harder to read.
+    ///
+    /// ### Known Problems
+    /// `.join()` introduces additional allocations that are not present when `PathBuf::push` is
+    /// used instead.
```

Comment:
Looking at this, I think it would be better to suggest `PathBuf` directly. Even if that might require additional modifications by the user. `PathBuf` is intended for modifications as this one AFAIK :)

Though, in that case we should have the applicability be less than machine applicable

Reply:
Sorry, to clarify, you mean you want to do some or all of the following?
1. Replace `.join()` with `.push()`
2. Replace the `Path::new` with `PathBuf::from`
3. Instead of suggesting a replacement, just flag it out to the user

Reply:
Sorry, if my comment wasn't clear. I mean a combination of the first two. So suggest `PathBuf::from()` and then add segments using `push()`. The thing is, that this probably change the type of the buffer and require further adjustments by the user. Therefore, we should display the suggestion, but not make it machine applicable.

Does this make sense, or should I try to describe it better? :)

Reply:
I'm looking at the docs for `PathBuf` now, and I just realised that there might be a recommended solution for situations like this :face_with_spiral_eyes:.

`[r"C:\", "windows", "system32.dll"].iter().collect::<PathBuf>()`

Will you be ok with this? Given it should still be machine applicable but I don't think it will have the same problems as before.

Reply:
You have to be careful when using the array as all the values have to be the same type.